sdk/src/cmd_disk.rs

```diff
@@ -0,0 +1,506 @@
+                            offset,
+                            base64_encoded_data,
+                        })
+                        .send()
+                        .await
+                        .map_err(OxideError::ApiError)?;
+
+                    pb.inc(data_len as u64);
+                }
+
+                Ok(())
+            }));
+
+            senders.push(tx);
+        }
+
+        // Read chunks from the file in the file system and send them to the
+        // upload threads.
+        let mut file = File::open(&disk.path).map_err(OxideError::IoError)?;
+        let mut i = 0;
+        let mut offset = 0;
+
+        let read_result: Result<(), _> = loop {
+            let mut chunk = Vec::with_capacity(CHUNK_SIZE as usize);
+
+            let n = match file.by_ref().take(CHUNK_SIZE).read_to_end(&mut chunk) {
+                Ok(n) => n,
+                Err(e) => {
+                    eprintln!("reading from {} failed with {:?}", disk.path, e,);
+                    break Err(OxideError::IoError(e));
+                }
+            };
+
+            if n == 0 {
+                break Ok(());
+            }
+
+            // If the chunk we just read is all zeroes, don't POST it.
+            if !chunk.iter().all(|x| *x == 0) {
+                let encoded = base64::engine::general_purpose::STANDARD.encode(&chunk[0..n]);
+
+                if let Err(e) = senders[i % UPLOAD_TASKS].send((offset, encoded, n)).await {
+                    eprintln!("sending chunk to thread failed with {:?}", e);
+                    break Err(OxideError::SendError(e));
+                }
+            } else {
+                // Bump the progress bar here to make it consistent
+                pb.inc(n as u64);
+            }
+
+            offset += CHUNK_SIZE;
+            i += 1;
+        };
+
+        for tx in senders {
+            drop(tx);
+        }
+
+        if let Err(e) = read_result {
+            // some part of reading from the disk and sending to the upload
+            // threads failed, so unwind. stop the bulk write process, finalize
+            // the disk, then delete it.
+            self.unwind_disk_bulk_write_stop().await?;
+            self.unwind_disk_finalize().await?;
+            self.unwind_disk_delete().await?;
+
+            // return the original error
+            return Err(e);
+        }
+
+        let mut results = Vec::with_capacity(handles.len());
+        for handle in handles {
+            let result = handle.await.map_err(OxideError::JoinError)?;
+            results.push(result);
+        }
+
+        if results.iter().any(|x| x.is_err()) {
+            // If any of the upload threads returned an error, unwind the disk.
+            eprintln!("one of the upload threads failed");
+            self.unwind_disk_bulk_write_stop().await?;
+            self.unwind_disk_finalize().await?;
+            self.unwind_disk_delete().await?;
+            return Err(OxideError::UploadThreadFailed);
+        }
```

Comment:
Perhaps something like:

```suggestion
        if let Some(err) = results.iter().filter(Result::is_err).next() {
            // If any of the upload threads returned an error, unwind the disk.
            eprintln!("one of the upload threads failed");
            self.unwind_disk_bulk_write_stop().await?;
            self.unwind_disk_finalize().await?;
            self.unwind_disk_delete().await?;
            return Err(err);
        }
```